src/stan_math_backend/Cpp.ml

```diff
@@ -0,0 +1,732 @@
+  let unused s =
+    [Comment "supress unused var warning"; Expression (Cast (Void, Var s))]
+end
+
+module Decls = struct
+  (** Declarations which get re-used often in the generated model *)
+
+  let current_statement =
+    VarDef
+      (make_var_defn ~type_:Int ~name:"current_statement__"
+         ~init:(Assignment (Literal "0")) () )
+
+  let dummy_var =
+    VarDef
+      (make_var_defn ~type_:Types.local_scalar ~name:"DUMMY_VAR__"
+         ~init:(Construction [Exprs.quiet_NaN])
+         () )
+    :: Stmts.unused "DUMMY_VAR__"
+
+  let serializer_in =
+    VarDef
+      (make_var_defn
+         ~type_:(TypeTrait ("stan::io::deserializer", [Types.local_scalar]))
+         ~name:"in__"
+         ~init:(Construction [Var "params_r__"; Var "params_i__"])
+         () )
+
+  let serializer_out =
+    VarDef
+      (make_var_defn
+         ~type_:(TypeTrait ("stan::io::serializer", [Types.local_scalar]))
+         ~name:"out__"
+         ~init:(Construction [Var "vars__"])
+         () )
+
+  let lp_accum t =
+    VarDef
+      (make_var_defn
+         ~type_:(TypeTrait ("stan::math::accumulator", [t]))
+         ~name:"lp_accum__" () )
+end
+
+type template_parameter =
+  | Typename of string  (** The name of a template typename *)
+  | RequireIs of string * string
+      (** A C++ type trait (e.g. is_arithmetic) and the template
+          name which needs to satisfy that.
+          These are collated into one require_all_t<> *)
+  | Require of string * string list
+  | Bool of string  (** A named boolean template type *)
+[@@deriving sexp]
+
+type cv_qualifiers = Const | Final | NoExcept [@@deriving sexp]
+
+type fun_defn =
+  { templates_init: template_parameter list list * bool [@default [[]], false]
+        (** Double list since some functions (mainly reduce_sum functors) need two sets of templates *)
+  ; name: identifier
+  ; inline: bool [@default false]
+  ; return_type: type_
+  ; args: (type_ * string) list
+  ; cv_qualifiers: cv_qualifiers list [@default []]
+  ; body: stmt list option }
+[@@deriving make, sexp]
+
+type constructor =
+  { args: (type_ * string) list
+  ; init_list: (identifier * expr list) list
+  ; body: stmt list }
+[@@deriving make, sexp]
+
+(** Incomplete list of C++ preprocessor directives *)
+type directive =
+  | Include of string
+  | IfNDef of string * defn list
+  | MacroApply of string * string list
+
+(** The Stan model class always has a non-default constructor and
+      destructor *)
+and class_defn =
+  { class_name: identifier
+  ; final: bool
+  ; base: type_
+  ; private_members: defn list
+  ; public_members: defn list
+  ; constructor: constructor
+  ; destructor_body: stmt list }
+
+and struct_defn =
+  {param: template_parameter option; struct_name: identifier; body: defn list}
+
+and defn =
+  | FunDef of fun_defn
+  | Class of class_defn
+  | Struct of struct_defn
+  | TopVarDef of var_defn
+  | TopComment of string
+  | TopUsing of string * type_ option
```

Comment:
Does top always just mean "At the top of the generated file?" Should we just call them global?

Reply:
It means "Top-level". I'm not married to the name Top, but I also think Global isn't the right one either (what is a "Global" comment, really?)

Reply:
I think a global comment is a thing, it's just a comment placed in global scope